### ISSUE TITLE
Support Novopen 6 / Echo Plus

### ISSUE THREAD

**Issue by madsviggers** [NightscoutFoundation/xDrip#1754]
The Novopen 6 is released in Denmark probably globally also. It works with NFC and should be build to work with all the apps that wants to use it. If needed I could buy one and send it to someone who could get it to work with xdrip?

**Comment by tolot27** [comment]
Thanks for the suggestion. I've contacted the German customer service and asked for the interface documentation. Unfortunately, the smartpens are not released in Germany, yet. Maybe, I can get two (6 and Echo Plus) for integration testing. If not, I'll contact you again.

**Comment by retroneo** [comment]
The interface documentation is here:

https://developer.novonordisk.com/home/register.html

**Comment by PieterDeBruijn** [comment]
I've made a request for the Echo Plus pen, so I'll let you know once it has arrived. I'm in The Netherlands. At least I can help test.

**Comment by PieterDeBruijn** [comment]
We will receive the new smart Echo Plus pen today/tomorrow (The Netherlands), so it's available over here. Let me know if I can test anything.

**Comment by Navid200** [comment]
Is this not similar to the one that is supported but isn't working now?
https://github.com/NightscoutFoundation/xDrip/issues/1176#issuecomment-894357275
I don't know if there is a point an adding more devices when we cannot support the ones that we already have.

**Comment by PieterDeBruijn** [comment]
I've had the message that it won't be delivered. No timeline as of yet. I guess we'll have to wait to start testing this one.

**Comment by madsviggers** [comment]
Have received the nocopen 6. Anything I can do?

**Comment by utzaki** [comment]
In Germany, it is available now: https://www.diashop.de/novopen-6-novopen-echo-plus

**Comment by tolot27** [comment]
My registration to the Novo Nordisk Developer Portal was still not processed. I've sent them a reminder.

**Comment by tolot27** [comment]
Maybe we do not get any response from Novo Nordisk because they require:

"NovoPen® 6 and NovoPen Echo® Plus are medical devices and therefore the integration of NovoPen® 6 and NovoPen Echo® Plus, and the use of insulin injection data, are subject to medical device regulations."

**Comment by yoshimo** [comment]
Which they could easily tell you. A no would be atleast an answer :(

**Comment by madsviggers** [comment]
![Screenshot_20211117_123630_com wakdev wdnfc](https://user-images.githubusercontent.com/16628542/142193747-a1250c18-9598-4a5a-8e8d-13d8eff30252.jpg)

**Comment by madsviggers** [comment]
Here is a screenshot from a nec reader tool. I have no idea if this is going to help, but this is what I can get from my novopen 6

**Comment by yoshimo** [comment]
The record might be easier to decode if you can find the raw hex values and some additional info that should be encoded in this value. Time&Date &Dose of the last injection if i remember the ad from novo nordisk well enough.

**Comment by utzaki** [comment]
According to https://www.novonordiskpro.de/diabetes/smartpens.html Diasend and Mysgr are already able to read the data. So maybe it is helpful to read values with NFC Tools and one of those two apps in parallel in order to have some samples.

**Comment by JtxGit** [comment]
I have also received such a device for testing and am busy playing with it using NFC on Android. I can read the UID and some data, but it does not seem that this data changes over time, so I don't think that this can be the data that we want... As a hex string it looks like:
e2000032800000000001002a5079002680000000800080000000000000000080000000080014650040016006400a0001010000000000

I wanted to try Diasend to see what the data looks like, but they require some account login... and I don't really feel like signing up just to test it.

**Comment by yoshimo** [comment]
> I wanted to try Diasend to see what the data looks like, but they require some account login... and I don't really feel like signing up just to test it.

You can look at their Android app. 
```

package com.diasend.diasend.nfc;
public class DeviceNovoPen implements DeviceExtractInterface {
    private static final int MAX_SINGLE_NDEF_READ_SIZE = 255;
    private static final int READ_RETRIES = 3;
    private static final long RETRY_WAIT_MILLIS = 50;
    private static final int device_transfer_done = 2;
    private static int maxMsgLen;
    private IsoDep mIsoDepTag;
    private ConnectivityNFC mStatusCallback;
    private static final byte[] NDEF_TAG_APP_SELECT = {0, -92, 4, 0, 7, -46, 118, 0, 0, -123, 1, 1, 0};
    private static final byte[] CAPABILITY_CONTAINER_SELECT = {0, -92, 0, 12, 2, -31, 3};
    private static final byte[] CAPABILITY_CONTAINER_READ = {0, -80, 0, 0, 15};
    private static final byte[] NDEF_FILE_SELECT = {0, -92, 0, 12, 2, -31, 4};
    private static final byte[] NDEF_EMPTY_FILE = {0, -42, 0, 0, 5, 0, 3, -48, 0, 0};
    private static final byte[] NDEF_READ_SIZE = {0, -80, 0, 0, 2};
    private static final byte[] PHD_EMPTY_RECORD = {-47, 3, 1, 80, 72, 68, 0, -112, 0};
    private boolean mTagIsInitialized = false;
    private final boolean mFollowTheSpec = true;
    private int mNfcMc = 0;
    private ByteArrayBuilder mOutputBuffer = null;
    private int mNumBytesReceived = 0;

    public DeviceNovoPen() {
        System.loadLibrary("A14037");
    }
```
The really intresting thing will happen in the native library shared object file `A14037`

**Comment by JtxGit** [comment]
Yes, I also looked at it, but have not found where and how they use the native library. I can send transceive commands without this library and I get replies, but this data that I get does not seem to change over time...

**Comment by utzaki** [comment]
My NovoPen 6 arrived today and I started testing it and looking at some documentation. Here are my findings:

- The chip seems to be this chip here: https://ams.com/documents/20143/36005/AS3953_PB000100_3-00.pdf/07e0ef0a-407f-e77a-acc9-dc929e3a1353
- It is using ISO 14443-4 which is a half-duplex transmission protocol that can be used for authentication purposes.

I'm afraid we're lost at this point. As long as we don't know more details about the authentication process, we won't be able to read the data. This may also be the reason why the above NDEF data are the same for every scan and do not include individual insulin treatment information.

**Comment by PaarnTh** [comment]
Only android app what i find to read my novopen echo plus is glooko.
Maybe someone can find authentication process from glooko app.

**Comment by tolot27** [comment]
Today, I got an answer from Novo Nordisk:

> We are sorry for the very long response time to your request.
> We have been in the process of developing our processes as we were making the developer portal available
> 
> Novo Nordisk has a strong focus on patient safety. For that reason, we strongly believe that our medical devices should be used together with digital solutions that are governed by regulatory authorities to ensure patient safety. While Nightscout is an open-source solution with an admirable purpose, but, as far as we know, it is not regulated by authorities. As such, Novo Nordisk cannot at this point support the technical integration with our medical devices. We will encourage Nightscout to engage with the regulatory authorities, and we are happy to further engage in discussions about integrations with our devices.
>
> Kind regards,
> The Novo Nordisk Digital Health Team

They are right to some extend. But providing the already injected amount of insulin to an app does not influence the patient safety at all. I'm still thinking about what to respond but don't believe that we will ever get access to the developer portal, unless we try to get some medical approval.

**Comment by tolot27** [comment]
I've closed the issue for now and tagged it with postponed. We can still discuss everything about integration here.

**Comment by Navid200** [comment]
https://github.com/NightscoutFoundation/xDrip/issues/2074
